index.bs

```diff
@@ -1370,77 +880,818 @@ What configuration is the device in after it resets?
+    1. Let |currAlternateInterface| be |interface|.{{USBInterface/alternates}}[|currAlternateIndex|].
+    1. <a>For each</a> |endpoint| of |currAlternateInterface|.{{USBAlternateInterface/endpoints}}:
+        1. Abort transfer currently scheduled on |endpoint| and <a>reject</a>
+            the associated promise with a {{AbortError}}.
+</div>
+
+<div algorithm="To find the interface index">
+    <dfn>To find the interface index</dfn> with the given |interfaceNumber| and |configuration| object,
+    performing the following steps:
+    1. Let |interfaceIndex| be 0.
+    1. While |interfaceIndex| smaller than the <a>size</a> of |configuration|.{{USBConfiguration/interfaces}}:
+        1. If |configuration|.{{USBConfiguration/[[interfaces]]}}[|interfaceIndex|].{{USBInterface/interfaceNumber}}
+            is equal to |interfaceNumber|, break.
+    1. Return |interfaceIndex|.
+</div>
+
+<div algorithm="To find the alternate index">
+    <dfn>To find the alternate index</dfn> with the given |alternateSetting| and |interface| object,
+    performing the following steps:
+    1. Let |alternateIndex| be 0.
+    1. While |alternateIndex| smaller than the <a>size</a> of |interface|.{{USBInterface/alternates}}:
+        1. If |interface|.{{USBInterface/[[alternates]]}}[|alternateIndex|].{{USBAlternateInterface/alternateSetting}}
+            is equal to |alternateSetting|, break.
+    1. Return |alternateIndex|.
+</div>
+
+<div algorithm="USBDevice constructor">
+    When a usb device is connected and detected, a new {{USBDevice}} object
+    representing a connected USB device is constructed by performing the following steps:
+    1. Set {{USBDevice/[[configurationValue]]}} to the returned value of <a>Get Configuration</a>.
+    1. Let |configurationDescriptors| be the result of
+        <a>to find a list of configuration descriptors for the connected usb device</a>.
+    1. <a>For each</a> |configurationDescriptor| of |configurationDescriptors|:
+        1. Let |configuration| be a <a>new</a> {{USBConfiguration}} object 
+            using <a>USBConfiguration(|device|,|configurationValue|)</a>
+            with |device| set to <a>this</a> and
+            |configurationValue| set to <code>bConfigurationValue</code> of |configurationDescriptor|.
+        1. <a>Append</a> |configuration| to {{USBDevice/[[configurations]]}}.
+        1. If <code>bConfigurationValue</code> of |configurationDescriptor| is equal to {{USBDevice/[[configurationValue]]}}:
+            1. Let |numInterfaces| be the <a>size</a> of |configuration|.{{USBConfiguration/interfaces}}.
+            1. Resize {{USBDevice/[[selectedAlternateSetting]]}} to |numInterfaces|.
+            1. Fill {{USBDevice/[[selectedAlternateSetting]]}} with 0.
+            1. Resize {{USBDevice/[[claimedInterface]]}} to |numInterfaces|.
+            1. Fill {{USBDevice/[[claimedInterface]]}} with <code>false</code>.
+</div>
+
+A device's <dfn>active configuration</dfn> is the combination of the
```

Comment:
Now that we have `[[configurationValue]]` and `[[selectedAlternateSetting]]` we don't need to have this definition for an "active configuration". We can define explicit algorithms for finding interfaces and endpoints in the active configuration like this step from `transferIn()`:

> Let endpoint be the IN endpoint in the [active configuration](https://pr-preview.s3.amazonaws.com/chengweih001/webusb/pull/212.html#active-configuration) with bEndpointAddress corresponding to endpointNumber. If there is no such endpoint [reject](https://webidl.spec.whatwg.org/#reject) promise with a [NotFoundError](https://webidl.spec.whatwg.org/#notfounderror) and abort these steps.

Reply:
I added new algorithm `<dfn>find the endpoint</dfn>` to be a helper for replacing the references using that `active configuration`

Updated in https://github.com/WICG/webusb/pull/212/commits/1a96f493fc285da99d2b812d809c241388197beb.